### PR TITLE
Add support for inheriting parameter estimates from a parent model

### PR DESCRIPTION
**Pull request by barrettk**
- adds new `inherit_param_estimates()` function
 - tests will fail on drone until nmrec PR is merged, a new release candidate is made, and bbr's drone file is updated to clone the latest nmrec release

closes https://github.com/metrumresearchgroup/bbr/issues/582

### CODE REVIEW

File: R/inherit-param-estimates.R [hunk 1]
```diff
@@ -0,0 +1,123 @@
+
+#' Supported parameter estimates to inherit from parent model
+BBR_ESTIMATES_INHERIT <- c("theta", "sigma", "omega")
```
Comment (seth127) [inline R/inherit-param-estimates.R:3]: Can you move this constant to `aaa.R` with the other constants?

Also, remove the `'` from the comment, because this is causing it render an `.Rd` file, which we don't need. After you remove that, please run `devtools::document()` and you should see that `.Rd` file disappear.

**Comment by seth127** [comment]
@kyleam [CI is failing](https://github-drone.metrumrg.com/metrumresearchgroup/bbr/3645/2/4) because (among other reasons) as of [ca57e37](https://github.com/metrumresearchgroup/bbr/pull/623/commits/ca57e378091c5d35b301c3be4940641727417654) we are requiring `cli >= 3.0.0` so that we can use `cli_abort()` and our "oldest" docker has an older version than that. "Oldest" is [currently set to 2020-06-08](https://github.com/metrumresearchgroup/bbr/blob/main/.drone.yml#L83), and it looks to me like we'd need to update to at least [2021-07-23](https://mpn.metworx.com/docs/packages/cli) to get `cli >= 3.0.0`.

What are your thoughts on that? We were just talking about possibly bumping our "oldest" to be more recent and this seems like a reasonably enough time and reason to do it. I wanted to get your thoughts first though.

Also, I can't remember if we actually have a `r-dev-ci-mpn-4.0:2021-07-23` docker image built, so we should obviously look at that too before we make the change.

**Comment by kyleam** [comment]
> @kyleam [CI is failing](https://github-drone.metrumrg.com/metrumresearchgroup/bbr/3645/2/4) because (among other reasons) as of [ca57e37](https://github.com/metrumresearchgroup/bbr/pull/623/commits/ca57e378091c5d35b301c3be4940641727417654) we are requiring `cli >= 3.0.0` so that we can use `cli_abort()` and our "oldest" docker has an older version than that. "Oldest" is [currently set to 2020-06-08](https://github.com/metrumresearchgroup/bbr/blob/main/.drone.yml#L83), and it looks to me like we'd need to update to at least [2021-07-23](https://mpn.metworx.com/docs/packages/cli) to get `cli >= 3.0.0`.

> What are your thoughts on that? We were just talking about possibly bumping our "oldest" to be more recent and this seems like a reasonably enough time and reason to do it.

I don't love the idea of bundling with this PR (mostly because it's nice to have a focused PR that presents the version decision and cleans up any now unnecessary kludges).  And I think it'd be easy enough to avoid the new cli patterns here, as the rest of this code base does.  But I agree we're due for a bump, so I don't have any strong objections if you want to do that here.

> Also, I can't remember if we actually have a `r-dev-ci-mpn-4.0:2021-07-23` docker image built, [...]

There is.

**Comment by barrettk** [comment]
@seth127 im requesting your review now, but there may still be something going on in drone. Re-running it now, but it failed to load the package for mpn-oldest. At the minimum everything else has been updated based on our other conversations

**Review by seth127 — CHANGES_REQUESTED** [review]
@barrettk I want to spend a bit more time on looking through the tests and playing with it myself, but I ran out of time today and thought it would be useful to put this part of the review in now. 

It's mostly docs updates. You can work on these and I'll continue the rest of my review in parallel. Thanks, and sorry for this being disjointed.

File: R/inherit-param-estimates.R [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+#'
+#' Set the initial parameter estimates of a model using the estimates of a
+#' previously executed model.
+#'
+#' @param .mod new model object to overwrite.
+#' @param .parent_mod Either a model object, or path to a model to inherit properties from.
+#' @param .inherit type of estimates to inherit from parent model.
+#' @param .bounds Whether to keep or discard the existing bounds when setting
+#'   the initial estimates in THETA records.
+#' @param .representation Whether to keep alternative representation options like
```
Comment (kyleam) [inline R/inherit-param-estimates.R:11]: Just a drive-by comment: This function is getting the values itself with `get_{theta,omega,sigma}` (i.e. it's the "final estimates from a previous NONMEM run" case from the parameter description), so it should always be "reset".  (`nmrec::set_{theta,omega,sigma}`, on the other hand, do not know the source of the values, so it seemed to me like it had to be up to the caller to decide the behavior.)

File: .drone.yml [hunk 3]
```diff
@@ -34,6 +34,9 @@ steps:
   image: 906087756158.dkr.ecr.us-east-1.amazonaws.com/r-dev-ci-mpn-4.0:latest
   commands:
   - R -s -e 'devtools::install_deps(upgrade = '"'"'never'"'"')'
+    # nmrec isn't available until the 2023-09-19 snapshot.
```
[the remark this answers is not in the document]
Comment (barrettk) [reply .drone.yml:37]: We can only remove the nmrec clone from MPN-latest after that time though right? Wondering if we want a separate comment for CRAN latest? I know we talked about cloning from the main branch in that case (similar to how `bbr.bayes` works I believe)

File: R/inherit-param-estimates.R [hunk 4]
```diff
@@ -0,0 +1,142 @@
+#' Inherit parameter estimates
+#'
+#' Set the initial parameter estimates of a model using the estimates of a
+#' previously executed model.
+#'
+#' @param .mod new model object to overwrite.
+#' @param .parent_mod Either a model object, or path to a model to inherit properties from.
```
[the remark this answers is not in the document]
Comment (barrettk) [reply R/inherit-param-estimates.R:7]: This is actually an important point - I dont think I tested out what would happen if multiple models were returned from `get_based_on`. I think I might need to add some extra handling for that

File: R/inherit-param-estimates.R [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+  return(.mod)
+}
+
+
+#' Confirm parent model path is valid, and error informatively if not
+#'
+#' @param .parent_mod Path to a parent model to inherit properties from.
+#'
+#' @keywords internal
+validate_parent_mod <- function(.parent_mod){
```
[the remark this answers is not in the document]
Comment (barrettk) [reply R/inherit-param-estimates.R:113]: I actually found that `rlang::abort()` can make use of the same syntax as `cli::cli_bullets()`. You cant format the coloring and formatting for code/italics like I had before, but you can have separate bullets. The 'x' creates an error message, and the 'i' creates an 'info' bullet. See below:

```r
> mod_est <- copy_model_from(MOD1, "mod_est", .overwrite = TRUE)
> mod_est2 <- copy_model_from(mod_est, "mod_est2", .overwrite = TRUE)

> inherit_param_estimates(mod_est2)
Error in `validate_parent_mod()` at bbr/R/inherit-param-estimates.R:72:5:
! Parent model (mod_est.ctl) exists, but has not been executed.
• To inherit parameter estimates from a parent model, `.parent_mod` must be a file path to a previously executed model.
Run `rlang::last_trace()` to see where the error occurred.
```

File: R/inherit-param-estimates.R [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+    parent_mod_path <- .parent_mod$absolute_model_path
+  }else{
+    parent_mod_path <- .parent_mod
+    # Confirm .parent_mod path is valid
+    validate_parent_mod(parent_mod_path)
+    .parent_mod <- read_model(parent_mod_path)
+  }
+
+  # Inherit model objects
+  .mod_path <- .mod$absolute_model_path
```
[the remark this answers is not in the document]
Comment (barrettk) [reply R/inherit-param-estimates.R:65]: I actually just realized that we should make sure this works with `.mod` extensions as well. I assume `nmrec::read_ctl` still works with that extension, but not sure if this is something we've thought about before

[the remark this answers is not in the document]
Comment (seth127) [reply R/inherit-param-estimates.R]: Yep, using the helper with make it support `.mod` too. See handling in [`find_nonmem_model_file_path`](https://github.com/metrumresearchgroup/bbr/blob/625df7acf74e254a5c0b5609ad4004bb0b520c86/R/get-path-from-object.R#L367) (used under the hood).

And, as far as I can tell, [`nmrec::read_ctl()` doesn't enforce any particular file extension](https://github.com/metrumresearchgroup/nmrec/blob/main/R/read-ctl.R#L34).

File: R/inherit-param-estimates.R (hunk 5 above)
[the remark this answers is not in the document]
Comment (seth127) [reply R/inherit-param-estimates.R:113]: That's nice. Thanks for the explanation.

**Review by seth127 — CHANGES_REQUESTED** [review]
Great work here @barrettk . I spent some more time with this today and I have some comments to clean up the test code, but overall this is very close. **I put all of my _new_ comments in a checklist in this comment**, instead of making them in-line, because this PR already has a lot of discussion on it. Hopefully that is helpful and not more confusing.

* [x] In [this comment](https://github.com/metrumresearchgroup/bbr/pull/623#discussion_r1425815744) about the `.inherit` default, you make the good point (that I missed): we actually aren't checking for multiple `based_on` models. Indeed, this currently causes a (weirdly obscure) failure. My suggestion:
  * Change change the default to `.inherit = get_model_on(.parent_mod)[1]`
  * Add this test (which fails without this ^ change):
    ```r
    it("works with multiple based_on models", {
      mod_est <- copy_model_from(MOD1, "mod_est", "Inherit estimates", .overwrite = TRUE) %>%
        add_based_on("../complex/acop-iov") %>%
        inherit_param_estimates()
      on.exit(delete_models(mod_est, .tags = NULL, .force = TRUE))
      
      expect_equal(length(get_based_on(mod_est)), 2L)

      mod1_params_final <- list(thetas = SUM1 %>% get_theta() %>% sprintf("%.3G", .))
      mod2_inits_inherit <- get_param_inits(mod_est)
      expect_equal(mod1_params_final$thetas, mod2_inits_inherit$thetas[[1]])
    })
    ```

* [x] I'm pretty sure [this `inherit_param_estimates()` call](https://github.com/metrumresearchgroup/bbr/blob/625df7acf74e254a5c0b5609ad4004bb0b520c86/tests/testthat/test-inherit-param-estimates.R#L76) can be deleted. The same thing is done two lines above.

* [x] Let's change `MOD1$absolute_model_path` to just `MOD1` in [this test expectation](https://github.com/metrumresearchgroup/bbr/blob/625df7acf74e254a5c0b5609ad4004bb0b520c86/tests/testthat/test-inherit-param-estimates.R#L95). It's a minor thing, but we're implicitly testing the "path" handling with the default case (because `get_based_on()` returns paths), so it would be good to test the "model" handling here.
  * Then you can also remove [these extra lines](https://github.com/metrumresearchgroup/bbr/blob/625df7acf74e254a5c0b5609ad4004bb0b520c86/tests/testthat/test-inherit-param-estimates.R#L110-L113) too. I'm happy to see that you thought of this, but this second `inherit_param_estimates()` call could essentially do nothing and this test would still pass (because you already inherited the same params above).

* [x] For this last test, [this comment about "only diagonals"](https://github.com/metrumresearchgroup/bbr/blob/625df7acf74e254a5c0b5609ad4004bb0b520c86/tests/testthat/test-inherit-param-estimates.R#L163) is actually _not_ true. I assume it just got copied from above, but modify that to reflect what we're actually doing here (and why).

* [x] I think we should avoid hard-coding any parameter values in the tests, because these are likely to change whenever we have to re-run the test models (which we do periodically). You've done a nice job of this already; I only see two places that would need to change:
  * [x] Testing that the bounds were inherited; you could change [this expectation](https://github.com/metrumresearchgroup/bbr/blob/625df7acf74e254a5c0b5609ad4004bb0b520c86/tests/testthat/test-inherit-param-estimates.R#L66-L69) to something like:
	```r
    expect_true(all(stringr::str_detect(
      mod2_params_inherit$thetas[[1]][1:3],
      "0, ?\\d"
    )))
    ```
  * [x] Testing inheriting only thetas; you could change these lines to something like:
    ```r
    mod1_inits <- get_param_inits(MOD1)
    mod2_inits_inherit <- get_param_inits(mod_est)

    # thetas changed
    expect_equal(mod1_params_final$thetas, mod2_inits_inherit$thetas[[1]])
    # Confirm omegas didnt change
    expect_equal(mod1_inits$omegas[[1]], mod2_inits_inherit$omegas[[1]])
    ```

* [ ] This `get_param_inits()` helper is nice for the tests, but I think the output can (and should) be simplified to return a list of vectors (instead of a list of lists). Looking at how you're using it and what you're comparing it to, I don't think you would ever want _multiple vectors_ for any given element (i.e. `thetas`). My suggestion:
  * Add an `unlist` to [this line](https://github.com/metrumresearchgroup/bbr/blob/4aac17b4fc1dd125795ff6a32daad4ef5955a127/tests/testthat/test-inherit-param-estimates.R#L34), like so
    ```r
    recs <- purrr::map(recs, function(.rec) unlist(purrr::map(.rec, extract_record_values)))
    ```
  * Remove the `[[1]]` from everywhere in the tests that you have things like `mod2_inits_inherit$thetas[[1]]`

* [ ] I agree with @kyleam that we shouldn't keep doing the "dot prefix" on new arguments for new functions. In this case, I would drop the leading dot from `.inherit`, `.bounds`, and `.digits`.

**Comment by barrettk** [comment]
@seth127 
>This get_param_inits() helper is nice for the tests, but I think the output can (and should) be simplified to return a list of vectors (instead of a list of lists). Looking at how you're using it and what you're comparing it to, I don't think you would ever want multiple vectors for any given element (i.e. thetas).

I changed everything else you mentioned so far except this. The main reason I did it this way was to make room for models that had _multiple_ OMEGA blocks and such if we ever needed to support that in the future. With your suggestion I think the returned results would get a little muddy for multiple records blocks. Granted - we dont currently test any models like that, so it could make the tests cleaner with our current models. Do you have any thoughts on this before I make the change?

**Comment by seth127** [comment]
> The main reason I did it this way was to make room for models that had multiple OMEGA blocks and such if we ever needed to support that in the future. With your suggestion I think the returned results would get a little muddy for multiple records blocks.

I'm not sure if it would get any muddier. I'm looking at how you use it, [here](https://github.com/metrumresearchgroup/bbr/blob/4aac17b4fc1dd125795ff6a32daad4ef5955a127/tests/testthat/test-inherit-param-estimates.R#L51-L62) for example: on line 61 you're comparing the result to what you got from `get_omegas()` (on line 54). Won't that be the same shape, no matter how many blocks were defined in the original control stream?

I guess, in general, I think once we're to the point where we're using this helper, the distinction of how many blocks the control stream had doesn't seem relevant anymore. Let me know if I'm missing something though.

**Comment by barrettk** [comment]
I was a little worried the concatenation of multiple record blocks would not line up with the returned results of `get_omega()` paired with `upper.tri` or `diag`. After looking into a bit more with an example model I think it's fine though, and we can adjust if we ever need to test one of those models


Using this model:
```r
mod <- read_model(file.path(MODEL_DIR_X, "example2_saemimp"))
```
Before:
```r
> mod_inits_inherit
$thetas
$thetas[[1]]
 [1] "0.7" "0.7" "2"   "2.0" "0.7" "0.7" "2.0" "2.0" "0.7" "0.7" "0.3"

$thetas[[2]]
[1] "4"


$omegas
$omegas[[1]]
 [1] "0.5"   "0.001" "0.5"   "0.001" "0.001" "0.5"   "0.001" "0.001" "0.001" "0.5"  

$omegas[[2]]
 [1] "0.01" "0.0"  "0.01" "0.0"  "0.0"  "0.01" "0.0"  "0.0"  "0.0"  "0.01"


$sigmas
$sigmas[[1]]
[1] "1.0"
```

After:

```r
> mod_inits_inherit
$thetas
 [1] "0.7" "0.7" "2"   "2.0" "0.7" "0.7" "2.0" "2.0" "0.7" "0.7" "0.3" "4"  

$omegas
 [1] "0.5"   "0.001" "0.5"   "0.001" "0.001" "0.5"   "0.001" "0.001" "0.001" "0.5"   "0.01"  "0.0"   "0.01"  "0.0"   "0.0"   "0.01"  "0.0"  
[18] "0.0"   "0.0"   "0.01" 

$sigmas
[1] "1.0"
```

**Review by seth127 — APPROVED** [review]
This is looking good. The only thing I see left is to add the extra clarifying sentence to the `.parent_mod` docstring (discussed in [this comment](https://github.com/metrumresearchgroup/bbr/pull/623#discussion_r1425815744)).

I'm approving anyway because that's not totally essential, but would be nice to add. Feel free to merge once that's done. Great job on this!
